src/ucs/sys/netlink.c

```diff
@@ -186,52 +212,108 @@ static ucs_status_t
+       present in the hash table before, initialize the list */
+    iface_rules = &kh_val(&ucs_netlink_routing_table_cache, iter);
+    if (khret != UCS_KH_PUT_KEY_PRESENT) {
+        ucs_list_head_init(iface_rules);
     }
 
-    if (ucs_bitwise_is_equal(ucs_sockaddr_get_inet_addr(info->sa_remote),
-                             dst_in_addr,
-                             ((const struct rtmsg *)nl_msg)->rtm_dst_len)) {
-        info->found = 1;
-        return UCS_OK;
+    /* allocate and initialize a new rule element */
+    new_rule = ucs_malloc(sizeof(route_entry_t), "route entry");
+    if (new_rule == NULL) {
+        ucs_error("could not allocate route entry");
+        return UCS_ERR_NO_MEMORY;
     }
 
+    new_rule->dest.ss_family = ((const struct rtmsg *)nl_msg)->rtm_family;
+    if (((const struct rtmsg *)nl_msg)->rtm_family == AF_INET) {
+        memcpy(&((struct sockaddr_in *)&new_rule->dest)->sin_addr,
+               dst_in_addr, sizeof(struct in_addr));
+    } else {
+        memcpy(&((struct sockaddr_in6 *)&new_rule->dest)->sin6_addr,
+               dst_in_addr, sizeof(struct in6_addr));
+    }
+
+    new_rule->if_index    = iface_index;
+    new_rule->subnet_mask = ((const struct rtmsg *)nl_msg)->rtm_dst_len;
+
+    /* add the new item to the list */
+    ucs_list_add_tail(iface_rules, &new_rule->list);
+
     return UCS_INPROGRESS;
 }
 
-int ucs_netlink_route_exists(const char *if_name,
-                             const struct sockaddr *sa_remote)
+static void ucs_netlink_lookup_route(ucs_netlink_route_info_t *info)
 {
-    ucs_netlink_route_info_t info = {0};
-    struct rtmsg rtm              = {0};
-    int iface_index;
+    ucs_list_link_t *iface_rules;
+    route_entry_t *curr_entry;
+    khiter_t iter;
+
+    iter = kh_get(ucs_netlink_rt_cache, &ucs_netlink_routing_table_cache,
+                  info->if_index);
+    if (iter == kh_end(&ucs_netlink_routing_table_cache)) {
+        return;
+    }
 
-    iface_index = if_nametoindex(if_name);
-    if (iface_index == 0) {
-        ucs_error("failed to get interface index (errno %d)", errno);
-        goto out;
+    iface_rules = &kh_val(&ucs_netlink_routing_table_cache, iter);
+    ucs_list_for_each(curr_entry, iface_rules, list) {
+        if (ucs_bitwise_is_equal(ucs_sockaddr_get_inet_addr(info->sa_remote),
+                                 ucs_sockaddr_get_inet_addr(
+                                    (const struct sockaddr *)&curr_entry->dest),
+                                 curr_entry->subnet_mask)) {
+            info->found = 1;
+            return;
+        }
+    }
+
+    return;
+}
+
+int ucs_netlink_route_exists(int if_index, const struct sockaddr *sa_remote)
+{
+    struct rtmsg rtm = {0};
+    ucs_netlink_route_info_t info;
+
+    pthread_mutex_lock(&ucs_netlink_rt_cache_lock);
+    if (route_data_exists) {
+        pthread_mutex_unlock(&ucs_netlink_rt_cache_lock);
+        goto lookup;
```

Comment:
1. Have `unlock` under `lookup`
2. Only `goto lookup` here

Then, `unlock` is only done in a single place

Reply:
instead of goto we can use UCS_INIT_ONCE